HPCC-20423 zapgen doesn't fail if credentials aren't provided against ldap enabled system

AttilaVamos opened the pull request:
Add code to handle curl return message:
- when authentication needed
- when authentication failed

Remove the generated ZAP file if it is empty (curl always generates outfile)
or authentication problem happens (it contains error information instead of
the real ZAP information)

Fix typo and add missing '\n' in passwordInput() fuction

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->
Tested manually in anuthentication enabled (10.173.160.101) and disabled
(local VM) system with and without credentials
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 6afec9c9aa192abf0a157d128846c79553f0a2a4
Build: success
Install hpccsystems-platform-community_7.0.0-rc2.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   35 sec (00:00:35) |  183 sec (00:03:03) |   94 sec (00:01:34) |   20 sec (00:00:20) |   42 sec (00:00:42) | 1543 sec (00:25:43) |   18 sec (00:00:18) | 1935 sec (00:32:15) |

AttilaVamos commented:
@cloLN @afishbeck Please review.

afishbeck commented on ecl/eclcmd/eclcmd_core.cpp:
```diff
@@ -1959,15 +1964,38 @@ class EclCmdZapGen : public EclCmdCommon
+            }
+            else if (strstr(errMsg.str(), "Authentication problem") != nullptr)
+            {
+                fprintf(stderr, "\nZapgen command returns with authentication problem. Please check your credentials.\n");
```
"returned with an"

afishbeck commented on ecl/eclcmd/eclcmd_core.cpp:
```diff
@@ -1959,15 +1964,38 @@ class EclCmdZapGen : public EclCmdCommon
-                return false;
+                isOk = false;
+            }
+            else if ((strstr(errMsg.str(), "< Location") != nullptr) && (strstr(errMsg.str(), "Login.html") != nullptr))
```
Would be nice if this were a more explicit error message being checked... is this output by ESP?  Is it the attempt to redirect to the login screen?

@wangkx is there a way for us to know that this is a REST client rather than someone that needs to go to the login page?

AttilaVamos replied:
Based on the curl report it seems a redirection attempt.

wangkx replied:
We do not have a way to detect a REST client. We are using the following rule to decide whether ESP should redirect a client to login page or simply send a 401:
bool rLoginPage = authorizationHeader.isEmpty() && !authReq.authBinding->isCORSRequest(originHeader.str()) &&
        !strieq(authReq.httpMethod.str(), POST_METHOD));

afishbeck replied:
So, 1. @AttilaVamos if the user had provided authentication are you adding it through CURL?
      2. @wangkx if auth is there it won't redirct?
      3. @AttilaVamos , can you send this as a POST instead of a GET?
      4. @wangkx if its a POST it won't redirect?

wangkx replied:
I am not sure that I understand the function here. For session enabled ESP, if a user is authenticated, ESP will not redirect to a login page. If a user has not authenticated, ESP will not redirect to the login page if the request is a POST.

afishbeck commented on ecl/eclcmd/eclcmd_core.cpp:
```diff
@@ -1959,15 +1964,38 @@ class EclCmdZapGen : public EclCmdCommon
+            Owned<IFile> file = createIFile(outputFile.str());
+            if (file->exists() && (removeOutFile || (0 == file->size())))
+                file->remove();
+            return false;
```
Return 0 above, false here... return type is int, should this be a number?

AttilaVamos replied:
Changed

afishbeck commented:
@AttilaVamos a few minor comments.

richardkchapman commented:
Checkboxes are missing

AttilaVamos commented:
@afishbeck Updated, please re-check. Based on Kevin's answer it seems we have not better way to handle login redirection.

HPCCSmoketest commented:
This PR (11714) is in the 2/3 position of the queue. Estimated start time is after ~1.0 hour(s)

HPCCSmoketest commented:
3/4. Process of PR-11714, label: HPCC-20423-fix-7.0.0 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 8c0619942b5f0ea5ba051b26d726e162fcec5456
Build: success
Install hpccsystems-platform-community_7.0.0-rc2.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   34 sec (00:00:34) |  185 sec (00:03:05) |   95 sec (00:01:35) |   22 sec (00:00:22) |   44 sec (00:00:44) | 1476 sec (00:24:36) |   18 sec (00:00:18) | 1874 sec (00:31:14) |

HPCCSmoketest commented:
1/1. Process of PR-11714, label: HPCC-20423-fix-7.0.0 starts now. Estimated completion time is ~ 0.5 hour(s)

AttilaVamos commented:
@wangkx @afishbeck updated, please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 8b633c0f404f18ec5b9cdc0050ffe2efa0884443
Build: success
Install hpccsystems-platform-community_7.0.0-rc2.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  187 sec (00:03:07) |   95 sec (00:01:35) |   21 sec (00:00:21) |   44 sec (00:00:44) | 1471 sec (00:24:31) |   21 sec (00:00:21) | 1848 sec (00:30:48) |

afishbeck commented on ecl/eclcmd/eclcmd_core.cpp:
```diff
@@ -1959,15 +1964,39 @@ class EclCmdZapGen : public EclCmdCommon
+            }
+
+            if (optVerbose)
                 fprintf(stderr, "Zapgen command returns with:\n%s\n", errMsg.str());
```
@AttilaVamos picky and not because of this PR, but should be "returned".

AttilaVamos replied:
Changed

afishbeck commented on ecl/eclcmd/eclcmd_core.cpp:
```diff
@@ -1959,15 +1964,39 @@ class EclCmdZapGen : public EclCmdCommon
-        return 0;
+        if (isOk)
+        {
+            fprintf(stdout, "ZAP file written into %s.\n", outputFile.str());
```
Again very picky, but I would so "to" rather than "into".

AttilaVamos replied:
Changed

afishbeck reviewed (changes requested):
@AttilaVamos a couple picky wording comments, other than that, looks good.

afishbeck commented:
@AttilaVamos Looks good, I like detecting the 401 error much better than the redirect page.

richardkchapman commented:
Please squash

AttilaVamos committed:
HPCC-20423 zapgen doesn't fail if credentials aren't provided against ldap enabled system

Add code to handle curl return message:
- when authentication needed
- when authentication failed

Remove the generated ZAP file if it is empty (curl always generates outfile)
or authentication problem happens (it contains error information instead of
the real ZAP information)

Fix typo and add missing '\n' in passwordInput() fuction

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>
